File: update_fontawesome_icons.py
#!/usr/bin/env python

# this file updates iconpicker.js to contain the most recent versions of icons.

import urllib.request
import yaml
import json
import re

# parse the yml file from font awesome
url = "https://raw.githubusercontent.com/FortAwesome/Font-Awesome/master/metadata/icons.yml"
f = urllib.request.urlopen(url)
myfile = f.read()
yml = yaml.load(myfile, Loader=yaml.FullLoader)

# create the icons json object, to be put into iconpicker.js
jsonf = {'icons':[]}
for d in yml: # loop through all keys of the yml dict
    for s in yml[d]['styles']: # create one object for each style (e.g. "brand", "solid") of an icon
      jsonf['icons'].append(
          {
          'title': 'fa'+s[0]+ ' fa-'+d, # create 'fab' class name for brand icons, "fas" for solid, etc.
          'searchTerms': yml[d]['search']['terms']
          }
      )
jsonf = json.dumps(jsonf, ensure_ascii=False) # prohibit ascii conversion, which breaks for special characters
f.close()

# replace icon json object inside iconpicker.js
with open('public_html/assets/js/iconpicker.js', 'r+') as f :
  jsf = f.read()
  jsf = re.sub(r'\{\n\s+icons(.*)\}',jsonf,jsf,flags=re.M|re.S)
  f.seek(0)
  f.write(jsf)
  f.truncate()

File: update_fontawesome_icons.php
<?php
//
// update_fontawesome_icons.json
// ---------------------------
// The pipeline_schema_builder.php page has an icon picker for FontAwesome icons
// so that pipeline authors can icon-ise their pipeline parameters.
// We load Font Awesome on the site through a JavaScript FA kit that always
// gives us the latest version of Font Awesome icons.
// This script runs on a cron job to update a JSON file describing the available
// FA icons so that the icon-picker has the full selection.

# parse the JSON file from Font Awesome
$json_url = 'https://raw.githubusercontent.com/FortAwesome/Font-Awesome/master/metadata/icons.json';
$raw_json = file_get_contents($json_url);
$fa_icons = json_decode($raw_json, true);

# create the icons json object, to be put into iconpicker.js
$icons = [];
foreach ($fa_icons as $d => $icon) {
    foreach ($icon['styles'] as $s) {
        # create one object for each style (e.g. "brand", "solid") of an icon
        $icon_class = 'fa' . $s[0] . ' fa-' . $d; # create 'fab' class name for brand icons, "fas" for solid, etc.
        $search_terms = implode(' ', $icon['search']['terms']);
        $icons[$icon_class] = $search_terms;
    }
}

# Write to a JSON file for the website
$results_fn = dirname(__FILE__) . '/public_html/assets/js/fa-icons.json';
$results_json = json_encode($icons, JSON_PRETTY_PRINT) . "\n";
file_put_contents($results_fn, $results_json);
